front/components/vaults/CreateVaultModal.tsx

```diff
@@ -0,0 +1,203 @@
+import { MinusIcon } from "lucide-react";
+import React, { useCallback, useContext, useMemo, useState } from "react";
+
+import { SendNotificationsContext } from "@app/components/sparkle/Notification";
+import logger from "@app/logger/logger";
+
+type RowData = {
+  icon: string;
+  name: string;
+  userId: string;
+  onClick?: () => void;
+  onMoreClick?: () => void;
+};
+
+type Info = CellContext<RowData, unknown>;
+
+interface CreateVaultModalProps {
+  owner: WorkspaceType;
+  isOpen: boolean;
+  onClose: () => void;
+  onSave: () => void;
+  allUsers: UserType[];
+}
+
+function getTableRows(allUsers: UserType[]): RowData[] {
+  return allUsers.map((user) => {
+    return {
+      icon: user.image ?? "",
+      name: user.fullName,
+      userId: user.sId,
+    };
+  });
+}
+
+function getTableColumns(
+  selectedMembers: string[],
+  handleMemberToggle: (member: string) => void
+): ColumnDef<RowData, unknown>[] {
+  return [
+    {
+      id: "name",
+      cell: (info: Info) => (
+        <DataTable.Cell avatarUrl={info.row.original.icon}>
+          {info.row.original.name}
+        </DataTable.Cell>
+      ),
+    },
+    {
+      id: "action",
+      cell: (info: Info) => {
+        const isSelected = selectedMembers.includes(info.row.original.userId);
+        return (
+          <Button
+            label={isSelected ? "Remove" : "Add"}
+            onClick={() => handleMemberToggle(info.row.original.userId)}
+            variant={isSelected ? "tertiary" : "secondary"}
+            size="sm"
+            icon={isSelected ? MinusIcon : PlusIcon}
+          />
+        );
+      },
+    },
+  ];
+}
+
+export function CreateVaultModal({
+  owner,
+  isOpen,
+  onClose,
+  onSave,
+  allUsers,
+}: CreateVaultModalProps) {
+  const [vaultName, setVaultName] = useState<string | null>(null);
+  const [selectedMembers, setSelectedMembers] = useState<string[]>([]);
+  const [searchTerm, setSearchTerm] = useState<string>("");
+  const [isSaving, setIsSaving] = useState(false);
+
+  const sendNotification = useContext(SendNotificationsContext);
+
+  const handleMemberToggle = useCallback((member: string) => {
+    setSelectedMembers((prevSelectedMembers) => {
+      const isCurrentlySelected = prevSelectedMembers.includes(member);
+      if (isCurrentlySelected) {
+        return prevSelectedMembers.filter((m) => m !== member);
+      } else {
+        return [...prevSelectedMembers, member];
+      }
+    });
+  }, []);
+
+  const createVault = async () => {
+    setIsSaving(true);
+    try {
+      const res = await fetch(`/api/w/${owner.sId}/vaults`, {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify({
+          name: vaultName,
+          members: selectedMembers,
+        }),
+      });
+
+      if (!res.ok) {
+        const errorData = await res.json();
+        throw new Error(errorData.error?.message || "Failed to create vault");
```

Comment:
This might happen if the vault's name is already taken, so we might want proper error handling?